File: 5 - Form Template Method/JS/CapitalStrategyTermLoan.js
import CapitalStrategy from "./CapitalStrategy";

export default class CapitalStrategyTermLoan extends CapitalStrategy{
    capital(loan) {
        return loan.getCommitment() * this.duration(loan) * this.riskFactor(loan)
    }

    duration(loan) {
        return this.weightedAverageDuration(loan);
    }

    weightedAverageDuration(loan) {
        let duration = 0
        const weightedAverage = loan.getPayments()
            .map(payment => this.yearsTo(payment.date, loan) * payment.amount)
            .reduce((a, b) => a + b)

        const sumOfPayments = loan.getPayments()
            .map(payment => payment.amount)
            .reduce((a, b) => a + b)

        if (this._commitment !== 0.0) {
            duration = weightedAverage / sumOfPayments
        }

        return duration
    }
}

File: 5 - Form Template Method/C#/MPG.FormTemplateMethod.Before/CapitalStrategyTermLoan.cs
﻿namespace MPG.FormTemplateMethod.Before
{
    using System;
    using System.Linq;

    public class CapitalStrategyTermLoan : CapitalStrategy
    {
        public override double Capital(Loan loan)
        {
            return loan.GetCommitment() * Duration(loan) * RiskFactor(loan);
        }

        public override double Duration(Loan loan)
        {
            return WeightedAverageDuration(loan);
        }

        private double WeightedAverageDuration(Loan loan)
        {
            var duration = 0.0;
            var weightedAverage = loan.GetPayments().Sum(payment => YearsTo(payment.Date, loan) * payment.Amount);
            var sumOfPayments = loan.GetPayments().Sum(payment => payment.Amount);

            if (Math.Abs(loan.GetCommitment()) > 0.0)
            {
                duration = weightedAverage / sumOfPayments;
            }

            return duration;
        }
    }
}